Added a SheetDismissSensitivity widget for dynamic configuration

bjartebore opened the pull request:
## Problem / Issue

#414 

## Solution

Will provide a way to configure the dismiss sensitivity level, this can be useful with sheets with dynamic height, where we want to set a dynamic value in userland

fujidaiti commented:
@bjartebore Thanks for the quick PR! Looks good at first glance. I'll review it properly and get back to you soon.

fujidaiti commented:
@bjartebore 

It seems like there are some misalignments between the proposed implementation and the spec described in #303.

In the PR, `minDragOffset` represents how much distance the user has to drag to close the modal sheet. However, the `dismissalOffset`, which is proposed in the spec and was called `minOffsetToDismiss` in the previous revision, represents a threshold in terms of `SheetOffset`, below which the sheet will be dismissed when the drag ends. (Note that the offset of a sheet is the distance from the bottom edge of the viewport to the top edge of the sheet, and the viewport is the rendering area of the modal route. Sorry for the unclear explanation in the spec.)

Roughly speaking, an ideal code would be like:
```dart
invokePop = resolvedDismissalOffset > currentOffset;
```

Where the current sheet offset can be obtained via `SheetViewportState.model.offset`. While this is straightforward, there is one thing that we must consider before calculating the `invokePop` value. It is the fact that 1. the offset stored in the `SheetViewportState.model` is a local value relative to the viewport and 2. the viewport moves downward with the sheet during the swipe down gesture. Because of those reasons, the apparent offset of the sheet may differ from the local offset. And the ideal behavior is that the sheet is dismissed when the offset is below the threshold as perceived by the user, rather than relative to the viewport. That's why we need to use the apparent sheet offset instead of the local offset. So the refined equation would be like:

```dart
final apparentOffset = currentOffset - viewportShift;
invokePop = resolvedDismissalOffset > apparentOffset;
```

Where the `viewportShift` is how many pixels the viewport moved until the drag ends, which is the same value as the `draggedDistance` in the existing code.

fujidaiti reviewed (changes requested):
It seems like there are some misalignments between the proposed implementation and the spec described in #303.

In the PR, `minDragOffset` represents how much distance the user has to drag to close the modal sheet. However, the `dismissalOffset`, which is proposed in the spec and was called `minOffsetToDismiss` in the previous revision, represents a threshold in terms of `SheetOffset`, below which the sheet will be dismissed when the drag ends. (Note that the offset of a sheet is the distance from the bottom edge of the viewport to the top edge of the sheet, and the viewport is the rendering area of the modal route. For example, if the sheet height is 100px and only half the sheet is visible, then the offset is 50px. Sorry for my unclear explanation in the spec.)

Roughly speaking, an ideal code would be like:
```dart
invokePop = resolvedDismissalOffset > currentOffset;
```

Where the current sheet offset can be obtained via `SheetViewportState.model.offset`. While this is straightforward, there is one thing that we must consider before calculating the `invokePop` value. It is the fact that 1. the offset stored in the `SheetViewportState.model` is a local value relative to the viewport and 2. the viewport moves downward with the sheet during the swipe down gesture. Because of those reasons, the apparent offset of the sheet may differ from the local offset. And the ideal behavior is that the sheet is dismissed when the offset is below the threshold as perceived by the user, rather than relative to the viewport. That's why we need to use the apparent sheet offset instead of the local offset. So the refined equation would be like:

```dart
final apparentOffset = currentOffset - viewportShift;
invokePop = resolvedDismissalOffset > apparentOffset;
```

Where the `viewportShift` is how many pixels the viewport moved until the drag ends, which is the same value as the `draggedDistance` in the existing code.

bjartebore commented:
Reading the text explains the text. You are right I missed the direction. I have adjusted the logic to fit the spec with your adjustments. I also updated the examples to use `SheetOffset.proportionalToViewport` as that makes more sense with this api.

bjartebore commented on lib/src/model.dart:
```diff
@@ -127,6 +132,29 @@ class ProportionalToViewportSheetOffset implements SheetOffset {
 
+/// A [SheetOffset] that evaluates an expression against [ViewportLayout].
+class ExpressionSheetOffset implements SheetOffset {
+  const ExpressionSheetOffset(this.expression);
```
Agree, not really a good use case for this either since SheetOffset already supports relative values, removed

bjartebore commented on test/modal_test.dart:
```diff
@@ -169,7 +169,7 @@ void main() {
             sensitivity: const SwipeDismissSensitivity(
               minFlingVelocityRatio: 1.0,
-              minDragDistance: 1000,
+              dismissalOffset: SheetOffset.absolute(1000),
```
changed

bjartebore commented on test/modal_test.dart:
```diff
@@ -198,7 +198,7 @@ void main() {
             sensitivity: const SwipeDismissSensitivity(
               minFlingVelocityRatio: 1.0,
-              minDragDistance: 1000,
+              dismissalOffset: SheetOffset.absolute(1000),
```
changed

bjartebore commented on lib/src/modal.dart:
```diff
@@ -566,12 +574,12 @@ class _SheetDismissibleState extends State<_SheetDismissible>
   const SwipeDismissSensitivity({
     this.minFlingVelocityRatio = 2.0,
-    this.minDragDistance = 200.0,
+    this.dismissalOffset = const SheetOffset.absolute(200),
```
changed
